src/mbgl/text/bidi.cpp

```diff
@@ -49,8 +58,86 @@ WritingDirection BiDi::baseWritingDirection(const std::u16string& input) {
     // the BiDi algorithm
     return ubidi_getBaseDirection(input.c_str(), static_cast<int32_t>(input.size())) == UBIDI_RTL
-        ? WritingDirection::RightToLeft
-        : WritingDirection::LeftToRight;
+    ? WritingDirection::RightToLeft
+    : WritingDirection::LeftToRight;
+}
+    
+// Takes UTF16 input in logical order and applies Arabic shaping to the input while maintaining logical order
+//  Output won't be intelligible until the bidirectional algorithm is applied
+std::u16string BiDi::applyArabicShaping(const std::u16string& input) {
+    UErrorCode errorCode = U_ZERO_ERROR;
+    
+    int32_t outputLength = u_shapeArabic(input.c_str(), static_cast<int32_t>(input.size()), NULL, 0,
+                                         (U_SHAPE_LETTERS_SHAPE&U_SHAPE_LETTERS_MASK)|(U_SHAPE_TEXT_DIRECTION_LOGICAL&U_SHAPE_TEXT_DIRECTION_MASK),
+                                         &errorCode);
+    
+    // Pre-flighting will always set U_BUFFER_OVERFLOW_ERROR
+    errorCode = U_ZERO_ERROR;
+    
+    std::unique_ptr<UChar[]> outputText = std::make_unique<UChar[]>(outputLength);
+    u_shapeArabic(input.c_str(), static_cast<int32_t>(input.size()), outputText.get(), outputLength,
+                  (U_SHAPE_LETTERS_SHAPE&U_SHAPE_LETTERS_MASK)|(U_SHAPE_TEXT_DIRECTION_LOGICAL&U_SHAPE_TEXT_DIRECTION_MASK),
+                  &errorCode);
+    
+    // If the algorithm fails for any reason, fall back to non-transformed text
+    if (U_FAILURE(errorCode))
+        return input;
+    
+    return std::u16string(outputText.get(), outputLength);
+}
+    
+std::u16string BiDi::applyBidiToParagraph(const std::u16string& input)
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+    
+    ubidi_setPara(bidiParagraph, input.c_str(), static_cast<int32_t>(input.size()), UBIDI_DEFAULT_LTR, NULL, &errorCode);
+
+    // TODO: Returning input string here is weird, since subsequent calls to setLine, etc. won't work. Maybe hold onto the errorCode so everything errors until setParagraph is called again?
+    if (U_FAILURE(errorCode))
+        return input;
+    
+    // Only valid as long as UBIDI_REMOVE_BIDI_CONTROLS and UBIDI_INSERT_LRM_FOR_NUMERIC are not set
+    int32_t outputLength = ubidi_getProcessedLength(bidiParagraph);
+    std::unique_ptr<UChar[]> outputText = std::make_unique<UChar[]>(outputLength);
+    ubidi_writeReordered(bidiParagraph, outputText.get(),outputLength, UBIDI_DO_MIRRORING, &errorCode);
+    if (U_FAILURE(errorCode))
+        return input;
+    
+    return std::u16string(outputText.get(), outputLength);
+}
+    
+std::u16string BiDi::getLine(int32_t start, int32_t end)
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+    ubidi_setLine(bidiParagraph, start, end, bidiLine, &errorCode);
+    
+    // TODO: At this point we're committed to the line breaking working. As a fallback we could try to just return the equivalent substring of the original reordered paragraph, but maybe we should just throw an exception?
```

Comment:
What kinds of errors could we expect to occur here? If it’s possible for line breaking to fail due to something in the vector tile data, then we might need a graceful fallback. Otherwise, if this is more about allocation or other catastrophic failures, throwing an exception would make more sense.